### PR TITLE
Harmonize CRS in the workflow

### LINKED ISSUE #297: Harmonize Coordinate Reference Systems (CRS) and set them as configuration options
## Checklist

- [x] I am using the current [`main`](https://github.com/pypsa-meets-africa/pypsa-africa/tree/main) branch or the latest [release](https://github.com/pypsa-meets-africa/pypsa-africa/releases). Please indicate.
- [x] I am running on an up-to-date [`pypsa-africa` environment](https://github.com/pypsa-meets-africa/pypsa-africa/blob/main/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`.

## Describe the Bug

Across all scripts, different Coordinate Reference Systems (CRS) have been used and there is confusion; they should be harmonized with criterion.
In particular, we need CRS for GIS analysis. Since we are interested into distances, EPSG reference type seems the most appropriate, [reference](https://docs.qgis.org/2.8/en/docs/gentle_gis_introduction/coordinate_reference_systems.html) and [reference 2](https://www.nceas.ucsb.edu/sites/default/files/2020-04/OverviewCoordinateReferenceSystems.pdf).
In particular, when it comes to CRS valid worldwide, EPSR:4326 may be the recommended and we are mainly using it, however it is in cylindrical reference (curved surface), thus it is not suitable when we need to measure distances in km.
For metric distances, we can use EPSG:3857, which seems to be the most appropriate.

Across the code, there are also different CRS that have been used and they shall be cleansed (refuse from PyPSA-Eur).
Moreover, I would like to have them as options, such as crs (EPSG:4326) and crs_metric (EPSG:3857) that can be customized.
Note that this does not mean to blindly substitute the optional value everywhere in the code, but where it is needed: when we set a crs on a loaded geodataframe from an external source (e.g. gadm) the crs may be specific to that external resource.
We shall specify the crs for our processed data.
That's why I some crs may be still hard coded, but it shall be discussed.

### PR DESCRIPTION
**Pull request by davide-f**
# Closes #297 

## Changes proposed in this Pull Request

This PR aims at harmonizing the CRSs used in the workflow as follows:
- Avoid having "magic" epsg specifications along the text: constant values are used in the scripts according to the file that are downloaded
- In the config files default and metric crs is proposed

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

File: config.tutorial.yaml [hunk 1]
```diff
@@ -91,6 +91,11 @@ enable:
   build_cutout: false
   build_natura_raster: true  # If True, then build_natura_raster can be run
 
+# definition of the Coordinate Reference Systems
+crs:
+  default_crs: EPSG:4326  # general geographic projection, not used for metric measures
+  metric_crs: EPSG:3085  # projection for metric measures only
```
Comment (pz-max) [inline config.tutorial.yaml]: REALLY COOL and well needed! One note:

There are projection that preserve angles, preserve areas and distances. Usually, there is not a one solution fits all case: https://learn.arcgis.com/en/projects/choose-the-right-projection/

Good for **preserving and calculating angles**:
- https://epsg.io/4326

Suitable for **preserving and calculating distances**:
- https://epsg.io/3857 aka. wgs 84 is the OpenStreetMap and Google Maps standard

Ok, for **preserving and caluclating areas**:
- we can go for https://epsg.io/3085 , but it's scope is for AMERICA. Maybe good

Comment (pz-max) [reply config.tutorial.yaml]: So maybe we should define this three options in the config.yaml?

Comment (davide-f) [reply config.tutorial.yaml]: Yeah, the 3085 was supposed to be 3857! A refuse, thanks!
Anyway, I wanted to add also some suggestions of alternative values

Comment (davide-f) [reply config.tutorial.yaml]: > So maybe we should define this three options in the config.yaml?

The 3085 should not be used unless the user focuses on America only.
The default values shall have global scope

File: scripts/build_renewable_profiles.py [hunk 2]
```diff
@@ -340,7 +345,7 @@
         else:
             availability = cutout.availabilitymatrix(regions, excluder, **kwargs)
 
-        area = cutout.grid.to_crs(3035).area / 1e6
+        area = cutout.grid.to_crs(metric_crs).area / 1e6
```
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: Atlite uses for areas 3035.
 https://github.com/PyPSA/atlite/blob/de20027a477b9ae79825fc2b9c5073a6f5a5f376/atlite/gis.py#L191-L206

File: scripts/build_renewable_profiles.py [hunk 3]
```diff
@@ -292,7 +297,7 @@
 
         capacity_per_sqkm = config["capacity_per_sqkm"]
 
-        excluder = atlite.ExclusionContainer(crs=3035, res=100)
+        excluder = atlite.ExclusionContainer(crs=metric_crs, res=100)
```
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: Should stay 3035. Atlite uses 3035 for exclusion stuff:  https://github.com/PyPSA/atlite/blob/de20027a477b9ae79825fc2b9c5073a6f5a5f376/atlite/gis.py#L191-L206

File: scripts/build_osm_network.py [hunk 4]
```diff
@@ -700,26 +702,29 @@ def fix_overpassing_lines(lines, buses, tol=1):
     df_to_add.set_index(lines.index[-1] + df_to_add.index, inplace=True)
 
     # update length
-    df_to_add["length"] = df_to_add.to_crs(epsg=3857).geometry.length
+    df_to_add["length"] = df_to_add.to_crs(metric_crs).geometry.length
```
Comment (pz-max) [inline scripts/build_osm_network.py]: Yes for line length (=distances) epsg 3857 is good

File: config.default.yaml [hunk 5]
```diff
@@ -35,6 +35,11 @@ enable:
   # requires cds API key https://cds.climate.copernicus.eu/api-how-to
   # More information https://atlite.readthedocs.io/en/latest/introduction.html#datasets
 
+# definition of the Coordinate Reference Systems
+crs:
+  default_crs: EPSG:4326  # general geographic projection, not used for metric measures. "EPSG:4326" is the standard used by OSM and google maps
```
Comment (pz-max) [inline config.default.yaml]: maybe something like:
```
crs:
   degree_crs : EPSG 4326  # (e.g. default for copernicus raster imports)
   distance_crs : EPSG:3857  # (e.g. OSM line lengths, distance from offshore..)
   area_crs : EPSG:3035  # (e.g. Atlite exclusion area container stuff)
```

File: scripts/build_renewable_profiles.py (hunk 2 above)
Comment (davide-f) [reply scripts/build_renewable_profiles.py]: Yes, but epsg:3035 is valid for europe: https://epsg.io/3035
That had to be modified

File: scripts/build_renewable_profiles.py (hunk 3 above)
Comment (davide-f) [reply scripts/build_renewable_profiles.py]: As above

File: scripts/build_osm_network.py (hunk 4 above)
Comment (davide-f) [reply scripts/build_osm_network.py]: The idea of metric_crs is to use a crs having as a unit meters (that's why metric but can be revised)

File: scripts/build_renewable_profiles.py (hunk 2 above)
Comment (pz-max) [reply scripts/build_renewable_profiles.py]: EPSG 3035 is however an EU crs... Let's maybe decide on an area crs & add something to the bug list for now? https://www.google.com/search?channel=fs&client=ubuntu&q=epsg+3035

Would be good to have an area crs depending on the region of interest

File: config.default.yaml (hunk 5 above)
Comment (davide-f) [reply config.default.yaml]: Revised, see next version

**Review by pz-max — APPROVED** [review]
Looks great. Ready to merge? @davide-f

**Comment by davide-f** [comment]
> Looks great. Ready to merge? @davide-f

Yes :)

**Pull request merged**